cmd/lotus-chainwatch/processor/processor.go

```diff
@@ -0,0 +1,301 @@
+		}
+
+		pts, err := p.node.ChainGetTipSet(ctx, types.NewTipSetKey(bh.Parents...))
+		if err != nil {
+			panic(err)
+		}
+
+		// collect all actors that had state changes between the blockheader parent-state and its grandparent-state.
+		// TODO: changes will contain deleted actors, this causes needless processing further down the pipeline, consider
+		// a separate strategy for deleted actors
+		changes, err = p.node.StateChangedActors(ctx, pts.ParentState(), bh.ParentStateRoot)
+		if err != nil {
+			panic(err)
+		}
+
+		// record the state of all actors that have changed
+		for a, act := range changes {
+			act := act
+			a := a
+
+			addr, err := address.NewFromString(a)
+			if err != nil {
+				panic(err)
+			}
+
+			ast, err := p.node.StateReadState(ctx, addr, pts.Key())
+			if err != nil {
+				panic(err)
+			}
+
+			// TODO look here for an empty state, maybe thats a sign the actor was deleted?
+
+			state, err := json.Marshal(ast.State)
+			if err != nil {
+				panic(err)
+			}
+
+			outMu.Lock()
+			if _, ok := actorsSeen[act.Head]; !ok {
+				_, ok := out[act.Code]
+				if !ok {
+					out[act.Code] = map[types.TipSetKey][]actorInfo{}
+				}
+				out[act.Code][pts.Key()] = append(out[act.Code][pts.Key()], actorInfo{
+					act:         act,
+					stateroot:   bh.ParentStateRoot,
+					height:      bh.Height,
+					tsKey:       pts.Key(),
+					parentTsKey: pts.Parents(),
+					addr:        addr,
+					state:       string(state),
+				})
+			}
+			actorsSeen[act.Head] = struct{}{}
+			outMu.Unlock()
+		}
+	})
+	return out, nil
+}
+
+func (p *Processor) unprocessedBlocks(ctx context.Context, batch int) (_ map[cid.Cid]*types.BlockHeader, err error) {
+	start := time.Now()
+	defer func() {
+		if err == nil {
+			log.Infow("Gathered Blocks to process", "duration", time.Since(start).String())
+		}
+	}()
+	rws, err := p.db.Query(`
+with toProcess as (
+    select blocks.cid, blocks.height, rank() over (order by height) as rnk
+    from blocks
+        left join blocks_synced bs on blocks.cid = bs.cid
+    where bs.is_processed is false and blocks.height != 0
+)
+select cid
+from toProcess
+where rnk <= $1
+`, batch)
+	if err != nil {
+		return nil, xerrors.Errorf("Failed to query for unprocessed blocks: %w", err)
+	}
+	out := map[cid.Cid]*types.BlockHeader{}
+
+	// TODO consider parallel execution here for getting the blocks from the api as is done in fetchMessages()
+	for rws.Next() {
+		var c string
+		if err := rws.Scan(&c); err != nil {
+			return nil, xerrors.Errorf("Failed to scan unprocessed blocks: %w", err)
+		}
+		ci, err := cid.Parse(c)
+		if err != nil {
+			return nil, xerrors.Errorf("Failed to parse unprocessed blocks: %w", err)
+		}
+		bh, err := p.node.ChainGetBlock(ctx, ci)
+		if err != nil {
+			// this is a pretty serious issue.
+			return nil, xerrors.Errorf("Failed to get block header %s: %w", ci.String(), err)
+		}
+		out[ci] = bh
+	}
```

Comment:
Calling rws.Next() might error. Make sure you check for it:
```
	if err := rws.Err(); err != nil {
		log.Fatal(err)
	}
```

Also.... rws? Was the `o` so expensive?